File: patterns/Creational/AbstractFactory/PHP/RealWorldExample/SystemUIComponents.php
<?php

/*
|--------------------------------------------------------------------------
| Abstract Factory Design Pattern - SystemUI Components
|--------------------------------------------------------------------------
| This example demonstrates the Factory Method Design Pattern to send various
| types of notifications (Email, SMS).
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/AbstractFactory/RealWorldExample
| @author    JawherKl
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/JawherKl/design-patterns-in-php
|--------------------------------------------------------------------------
*/

namespace Creational\AbstractFactory\RealWorldExample;

/**
 * Abstract Factory for creating UI components.
 */

interface UIFactory
{
    public function createButton(): Button;
    public function createCheckbox(): Checkbox;
}

/**
 * Concrete Factory for Windows OS.
 */
class WindowsFactory implements UIFactory
{
    public function createButton(): Button
    {
        return new WindowsButton();
    }

    public function createCheckbox(): Checkbox
    {
        return new WindowsCheckbox();
    }
}

/**
 * Concrete Factory for Mac OS.
 */
class MacFactory implements UIFactory
{
    public function createButton(): Button
    {
        return new MacButton();
    }

    public function createCheckbox(): Checkbox
    {
        return new MacCheckbox();
    }
}

/**
 * Abstract product interface for buttons.
 */
interface Button
{
    public function render(): string;
}

/**
 * Concrete product for Windows buttons.
 */
class WindowsButton implements Button
{
    public function render(): string
    {
        return "Rendering Windows button.";
    }
}

/**
 * Concrete product for Mac buttons.
 */
class MacButton implements Button
{
    public function render(): string
    {
        return "Rendering Mac button.";
    }
}

/**
 * Abstract product interface for checkboxes.
 */
interface Checkbox
{
    public function toggle(): string;
}

/**
 * Concrete product for Windows checkboxes.
 */
class WindowsCheckbox implements Checkbox
{
    public function toggle(): string
    {
        return "Toggling Windows checkbox.";
    }
}

/**
 * Concrete product for Mac checkboxes.
 */
class MacCheckbox implements Checkbox
{
    public function toggle(): string
    {
        return "Toggling Mac checkbox.";
    }
}

/**
 * Client code.
 */
function renderUI(UIFactory $factory)
{
    $button = $factory->createButton();
    echo $button->render();

    $checkbox = $factory->createCheckbox();
    echo $checkbox->toggle();
}

echo "Testing Windows UI factory:\n";
renderUI(new WindowsFactory());
echo "\n\n";

echo "Testing Mac UI factory:\n";
renderUI(new MacFactory());

File: patterns/Creational/AbstractFactory/Go/RealWorldExample/SystemUIComponents.go
package main

import "fmt"

// UIFactory - Abstract Factory for creating UI components
type UIFactory interface {
    CreateButton() Button
    CreateCheckbox() Checkbox
}

// WindowsFactory - Concrete Factory for Windows OS
type WindowsFactory struct{}

func (w *WindowsFactory) CreateButton() Button {
    return &WindowsButton{}
}

func (w *WindowsFactory) CreateCheckbox() Checkbox {
    return &WindowsCheckbox{}
}

// MacFactory - Concrete Factory for Mac OS
type MacFactory struct{}

func (m *MacFactory) CreateButton() Button {
    return &MacButton{}
}

func (m *MacFactory) CreateCheckbox() Checkbox {
    return &MacCheckbox{}
}

// Button - Abstract product interface for buttons
type Button interface {
    Render() string
}

// WindowsButton - Concrete product for Windows button
type WindowsButton struct{}

func (w *WindowsButton) Render() string {
    return "Rendering Windows button."
}

// MacButton - Concrete product for Mac button
type MacButton struct{}

func (m *MacButton) Render() string {
    return "Rendering Mac button."
}

// Checkbox - Abstract product interface for checkboxes
type Checkbox interface {
    Toggle() string
}

// WindowsCheckbox - Concrete product for Windows checkbox
type WindowsCheckbox struct{}

func (w *WindowsCheckbox) Toggle() string {
    return "Toggling Windows checkbox."
}

// MacCheckbox - Concrete product for Mac checkbox
type MacCheckbox struct{}

func (m *MacCheckbox) Toggle() string {
    return "Toggling Mac checkbox."
}

// Client function to render UI components
func renderUI(factory UIFactory) {
    button := factory.CreateButton()
    fmt.Println(button.Render())

    checkbox := factory.CreateCheckbox()
    fmt.Println(checkbox.Toggle())
}

func main() {
    fmt.Println("Testing Windows UI factory:")
    renderUI(&WindowsFactory{})
    fmt.Println()

    fmt.Println("Testing Mac UI factory:")
    renderUI(&MacFactory{})
}
